fixes requireJS mixins racecondition with require calls in dom

AndresInSpace opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Fixes the race condition with DOM require calls parsing before `require(['mixins'],{...});` replacing require function.
Resolves mixins not applying.
Ensures jquery-mixin from security https://github.com/magento/security-package/blob/develop/ReCaptchaWebapiUi/view/frontend/requirejs-config.js#L10 is mixed in before jquery called

**How to Confirm You Have This Issue:**
add to custom block output in top of page of cart closest to body opening tag, visit empty cart page `/checkout/cart/index`
context should be from wrapper (mixin applied) not from customer-data (mixin not applied)
```
<script>
    require(['jquery', 'Magento_Customer/js/customer-data'], function($, customerData) {
        'use strict';

        console.log("context from mixin?: %O", customerData.getExpiredSectionNames);
    });
</script>
```

Original Process:
requirejs makeRequire calls nexTick (setTimeout) freeing thread to parse DOM
requirejs/mixins.js loads and define call is made
DOM require calls race condition - if calls happen require context is from requirejs/require.js and not mixins.js
requirejs/mixins.js require fired and now instantiating and replacing require call context to mixins.js

Changes:
Moves requirejs-config.js before mixins.js ensures configurations are loaded
Remove context.nextTick wrap from makeRequire return localRequire function (this was freeing thread to DOM) **
Remove redundant intakeDefines();
Move require calls out of require-configs. Only one is for jquery no conflict. Remove from https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Theme/view/base/requirejs-config.js#L87 and move to end of  `require(['mixins], function(..) { ... {beforeClosingtag} });` ensures mixin replaces require function so jquery-mixin is properly applied

**`checkLoaded()` has all the setTimeout functionality we need https://github.com/magento/magento2/blob/9124539dd5d26ea31dcda8608071582d2bfbde1a/lib/web/requirejs/require.js#L717 
See more information in my comment on the issue this PR should fix below.
### Related Pull Requests
<!-- related pull request placeholder -->
### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#33593
2. Properly Fixes magento/magento2#31920
3. Addresses #38274 (config output may need hoisting see my comment in ticket)
4. Should also fix magento/baler#23

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
See the above issue for steps to reproduce mixins not loading issue caused from this race condition
additionally can
1. increase nextTick setTimeout value in requirejs.js to a larger number,increment more if needed to replicate consistently
5. reload page 
6. verify dom require calls are parsed before require(['mixins']

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

AndresInSpace committed:
output requirejs-config.js before requirejs/mixins.js

AndresInSpace committed:
remove nextTick from makeRequire to stop freeing thread ensure mixins loads before dom parse

AndresInSpace committed:
remove require call setting jquery no conflict from require config block

AndresInSpace committed:
add require jquery no conflict to require mixins - ensure recaptchawebapiui jquery-mixin load before require calls in dom

AndresInSpace commented:
@magento give me test instance

On Thu, Aug 22, 2024, 10:16 PM m2-assistant[bot] ***@***.***>
wrote:

> Hi @AndresInSpace <https://github.com/AndresInSpace>. Thank you for your
> contribution!
> Here are some useful tips on how you can test your changes using Magento
> test environment.
> Add the comment under your pull request to deploy test or vanilla Magento
> instance:
>
>    - @magento give me test instance - deploy test instance based on PR
>    changes
>    - @magento give me 2.4-develop instance - deploy vanilla Magento
>    instance
>
> ❗ Automated tests can be triggered manually with an appropriate comment:
>
>    - @magento run all tests - run or re-run all required tests against
>    the PR changes
>    - @magento run <test-build(s)> - run or re-run specific test build(s)
>    For example: @magento run Unit Tests
>
> <test-build(s)> is a comma-separated list of build names.
> Allowed build names are:
>
>    1. Database Compare
>    2. Functional Tests CE
>    3. Functional Tests EE
>    4. Functional Tests B2B
>    5. Integration Tests
>    6. Magento Health Index
>    7. Sample Data Tests CE
>    8. Sample Data Tests EE
>    9. Sample Data Tests B2B
>    10. Static Tests
>    11. Unit Tests
>    12. WebAPI Tests
>    13. Semantic Version Checker
>
> You can find more information about the builds here
> <https://github.com/magento/magento2/wiki/Magento-Automated-Testing>
> ℹ️ Run only required test builds during development. Run all test builds
> before sending your pull request for review.
> ------------------------------
>
> For more details, review the Code Contributions
> <https://developer.adobe.com/commerce/contributor/guides/code-contributions/>
> documentation.
> Join Magento Community Engineering Slack
> <https://developer.adobe.com/open/magento/slack> and ask your questions
> in #github <https://magentocommeng.slack.com/messages/C7KB93M32> channel.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/magento/magento2/pull/39097#issuecomment-2306059135>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADNQ47QTMLXF2NWF4BSYBDDZS2LQDAVCNFSM6AAAAABM7IWNX6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMBWGA2TSMJTGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

AndresInSpace commented:
@magento give me 2.4-develop instance

AndresInSpace commented:
Appears resolved based on testing customer-data in a require statement added via home block so it is very close to opening body tag in order to trigger the error.

View console log F12 & Reload w cache disabled or ctrl+f5 hard refresh chrome on windows
Vanilla Instance: https://45243a10eb203fde7e8eedf63293cfe1.instances-prod.magento-community.engineering/
PR Instance: https://0a05fa17a7983d4b098002b067de8a5d.instances-prod.magento-community.engineering/

AndresInSpace commented:
@magento run all tests

AndresInSpace commented:
@magento give me test instance

ihor-sviziev commented on lib/web/requirejs/require.js:
```diff
@@ -1445,25 +1445,18 @@ var requirejs, require, define;
                     //Grab defines waiting in the global queue.
                     intakeDefines();
 
-                    //Mark all the dependencies as needing to be loaded.
```
@AndresInSpace, since this is the change in RequireJs lib, I think you should create this pull request in https://github.com/requirejs/requirejs. And then just update the lib here

AndresInSpace replied:
@ihor-sviziev I don't believe it belong as a native change to requirejs because in a vanilla state, requirejs won't have the require function replaced. So in vanilla requirejs it works as is.
This is a Magento fix due to magento mixins.js replacing require method IMO.

Should we move it from `lib/web/requirejs/require.js` and into `lib/web/mage/requirejs/require.js` then since it will be a modified library specific to magento?

AndresInSpace replied:
https://github.com/requirejs/requirejs/pull/1859

AndresInSpace commented:
@magento give me test instance

AndresInSpace commented:
@magento give me 2.4-develop instance

AndresInSpace commented:
@magento give me 2.4-develop instance

AndresInSpace commented:
@magento give me test instance

AndresInSpace committed:
Update require.js lib

based on acceptance of requirejs/requirejs/pull/1859 removes nextTick cause of race condition

fabiopelloso commented:
https://github.com/magento give me 2.4.7 instance with extensions adyen/module-payment:9.5.3

ioweb-gr commented:
Hi everyone, any ideas of when this will be merged? We've been suffering lately due to this with the page builder module and generally mixins not always loading preventing saving cms blocks and pages. 

Applying this patch temporarily seems to solve the issues.

medigeek commented:
@Den4ik you self-requested a review [2 months ago](https://github.com/magento/magento2/pull/39097#event-14302441009) -- is there anything left to be reviewed to push this fix forward? Thank you.

Den4ik reviewed (changes requested):
Hi @AndresInSpace 
Could you please resolve merge conflict

ioweb-gr commented:
Hi @AndresInSpace,

Could you please recheck this PR and help resolve the merge conflicts when you have a moment? It's highly anticipated and would be a valuable addition to the community!

Thank you for your efforts :)

medigeek commented:
@Den4ik Hi, I updated instead of @AndresInSpace and fixed the conflicts in a fork:
https://github.com/magento/magento2/compare/2.4-develop...medigeek:magento2:pr-39097-rebased-fixed-conflicts-jan-2025
https://github.com/medigeek/magento2/tree/pr-39097-rebased-fixed-conflicts-jan-2025

most stuff in require.js have already been fixed previously, but the other files weren't.
hope it helps.

Den4ik commented:
Hi @medigeek 
Could you please merge your branch into `AndresInSpace:requirejsfix` to proceed with current PR.

ioweb-gr commented:
I believe without @AndresInSpace help that's not gonna happen @Den4ik . Perhaps we should process as a new PR from @medigeek 's branch?

medigeek commented:
I had to remove the branch to see I can't do what was asked... now i have to re-do the changes 🤣

ioweb-gr commented:
@medigeek @Den4ik  I tried to do it myself, if you guys want to check #39691 if everything is in place we can process that one.

davidwindell commented:
Unfortunately this (and #39691) doesn't solve the issue outlined in #38274 for at least three sites we manage. The only thing that reliable does (without any other patches) is the diff mentioned in [this comment](https://github.com/magento/magento2/issues/38274#issuecomment-1913186542) i.e. https://github.com/magento/magento2/commit/847c7997143d3c90b26f94480a9fe202f499faf7.diff

ioweb-gr commented:
@davidwindell It might not solve that exact issue because the reason might be different but it solves a ton of other issues with race conditions in mixins preventing them from loading well. 

For example 

- https://github.com/magento/magento2/issues/35097 

Which was closed because there was no easy way to reproduce. 

However as soon as we apply this patch, the error there is gone. Latest person who faced it can confirm it @jorgehs91 

There are other scattered issues around about this thing with mixins not being loaded reliably. 

I don't believe the fact that it doesn't solve #38274  as you mentioned should prevent this from getting merged imho.

amine-y commented:
We are facing the same issue in magento 2.4.7-p4

amine-y commented:
@ihor-sviziev Is there any way to speed up the process of this PR, please? So many people are experiencing the same issue.

ioweb-gr commented:
#39691 We are trying to reprocess this PR here because this one is stale but I can see some of the tests are failing.

We need to check what changes need to be done so that tests stop failing if we want a shot at pushing this fast

ioweb-gr commented:
@Den4ik unfortunately the patch in the PR doesn't entirely fix the issue. I've encountered installations where the patch half-works so I closed the PR. This is a major issue affecting customers randomly. 

I've been collecting errors with sentry from various such cases. 

![image](https://github.com/user-attachments/assets/fd2db86f-55ee-4b4a-a26e-9938e246ac76)

Randomly files that exist in the file system are not being loaded

Den4ik commented:
Hi @ioweb-gr 
Thanks for your message. Could you please provide little bit more information that can help idenrify core of issue?

ioweb-gr commented:
This issue manifests the same as Issue #33593.

From the errors I've observed, it seems that the aliases defined in requirejs-config.js fail to load properly at random as explained by @AndresInSpace 


For example:

require(['<component>'], function(component) {
    // This sometimes fails
});

However, using the full Magento asset path consistently works:

require(['Magento_Ui/js/<component>'], function(component) {
    // This always works
});

This suggests that the issue lies with the inconsistent resolution of aliases with a possible race condition 

This can manifest in the frontend as inability to load custom options or add to cart button or the checkout page randomly

In the backend it prevents saving page builder pages or wysiwyg content randomly. 


Unfortunately I didn't find the root cause

The solution in the PR helped in a few clients but failed in a couple of them aggravating the issue ever further

engcom-Hotel reviewed (commented):
Hello @AndresInSpace,

Thanks for the contribution!

Please resolve the conflicts on this PR, so that we can proceed with this. Meanwhile moving this PR `On Hold`.

Thanks

ioweb-gr commented:
@engcom-Hotel I resolved the merge conflicts in another PR, but this doesn't seem to fix everything properly for all cases. I stumbled into a case where applying this patch actually made things worse in a website. So it's not a matter of just resolving the merge conflicts.

engcom-Hotel commented:
Thanks @ioweb-gr for your inputs.

Can you please provide us the PR on which you have resolved the conflicts?

Thanks
